include/oneapi/dpl/pstl/iterator_impl.h

```diff
@@ -401,7 +401,10 @@ class transform_iterator
 {
   private:
     _Iter __my_it_;
-    const _UnaryFunc __my_unary_func_;
+    _UnaryFunc __my_unary_func_;
+
+    static_assert(std::is_invocable_v<const _UnaryFunc, typename std::iterator_traits<_Iter>::reference>,
+                  "Unary Functor is not const callable with the reference type of base iterator.");
```

Comment:
Ma be better to use type name in this message?
`"_UnaryFunc is not const callable with the reference type of base iterator."`

Reply:
Why you are using `const _UnaryFunc` instead of `_UnaryFunc` here?
Because you already declared it as
`_UnaryFunc __my_unary_func_;` ...

Reply:
It is because we want to confirm that `_UnaryFunc` is callable from within `operator*` of `transform_iterator` which is marked as `const` (meaning that member fields cannot be modified within the body).  This check is to ensure that a `_UnaryFunc` has a `const` call operator which would be eligible to be called during a dereference.

Reply:
If it's needed, we able to avoid this restriction, I think...

Reply:
> May be better to use `decltype(*__my_it_)` here instead of typename std::iterator_traits<_Iter>::reference` ?

Are you suggesting this just to shorten the number of characters?  It seems like these should be the same as long as `reference` is defined correctly.  I suppose using `decltype(*__my_it_)` removes any question of that, but it seems perhaps less clear / readable.

Reply:
> If it's needed, we able to avoid this restriction, I think...

I don't think I understand, are you saying that this `static_assert` is unnecessary because we already would get a compile time error from it not being there?  
This is true, but as mentioned in this thread, the difference is:
1) better error message for the user 
2) check on instantiation, without requiring an attempted dereference.  (This is perhaps not an advantage, but it is a difference)

Reply:
> Ma be better to use type name in this message? `"_UnaryFunc is not const callable with the reference type of base iterator."`

Done.

Reply:
> > If it's needed, we able to avoid this restriction, I think...
> 
> I don't think I understand, are you saying that this `static_assert` is unnecessary because we already would get a compile time error from it not being there? This is true, but as mentioned in this thread, the difference is:
> 
> 1. better error message for the user
> 2. check on instantiation, without requiring an attempted dereference.  (This is perhaps not an advantage, but it is a difference)

I mean that we can make local copy of this `__my_unary_func_ ` if `std::is_invocable_v<const _UnaryFunc, typename std::iterator_traits<_Iter>::reference>` condition isn't true.

Reply:
> I mean that we can make local copy of `__my_unary_func_ ` if `std::is_invocable_v<const _UnaryFunc, typename std::iterator_traits<_Iter>::reference>` condition isn't true.

I would argue that we specifically do not want to give the impression that we support such functors which are not const callable.  My understanding of our initial reason for making the functor itself `const` was to portray this idea that the functor should be stateless and without side effects.  I see this `static_assert` as helping to help maintain clarity in our portrayal of this expectation for the functors.

Practically, we could support such functors in the way you describe, but it wouldn't *actually* support any extra functionality which is allowed by having a non-const `operator()`.  (It also would add some requirements on copy constructability of the functors.)

Reply:
I understood, ok.